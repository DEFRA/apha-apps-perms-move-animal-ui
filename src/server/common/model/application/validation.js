/**
 * @import {SectionValidationResult} from '../section/validation.js'
 * @import {SectionModel} from '../section/section-model/section-model.js'
 */

import mapValues from 'lodash/mapValues.js'

/**
 * @typedef {{[key:string]: SectionValidationResult}} ValidationResult
 * @typedef {{
 *   isValid: boolean,
 *   result: ValidationResult
 * }} ApplicationValidationResult
 */

/**
 * Validates all the answers within the application and outputs an overall isValid result.
 * @param {{[key:string]: SectionModel}} data - The data to be validated.
 * @returns {ApplicationValidationResult} An object containing the validation result
 */
export const validateApplication = (data) => {
  const result = mapValues(data, (answer) => answer.validate())
  const isValid = Object.values(result).every((a) => a.isValid)

  return {
    isValid,
    result
  }
}
