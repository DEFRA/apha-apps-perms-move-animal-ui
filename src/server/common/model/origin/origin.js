import Joi from 'joi'
import { Model, validateAgainstSchema } from '../model.js'
import { OnOffFarm } from './on-off-farm.js'
import { CphNumber } from './cph-number.js'
import { Address } from './address.js'

const validationSchema = Joi.object().custom((value, helpers) => {
  const invalid = Object.values(value).some((item) => {
    return !item.validate().isValid
  })

  if (invalid) {
    return helpers.error('any.invalid')
  }
  return value
})

/**
 * export @typedef {{
 *  onOffFarm: OnOffFarmData | undefined;
 *  cphNumber: CphNumberData | undefined;
 *  address: AddressData | undefined;
 * }} OriginData
 *
 * export @typedef {{
 *  onOffFarm?: OnOffFarm;
 *  cphNumber: CphNumber;
 *  address?: Address;
 * }} OriginDataObjects
 *
 * @import {OnOffFarmData} from './on-off-farm.js'
 * @import {CphNumberData} from './cph-number.js'
 * @import {AddressData} from './address.js'
 * @import {RawPayload} from '../model.js'
 */

export class Origin extends Model {
  /**
   * @returns {RawPayload | undefined}
   */
  get value() {
    return this._data
  }

  /**
   *
   * @returns {RawPayload | undefined}
   */
  toState() {
    return this.value
  }

  validate() {
    return validateAgainstSchema(validationSchema, this._data)
  }

  /**
   * @param {OriginData | undefined} state
   * @returns {Origin}
   */
  static fromState(state) {
    return new Origin({
      onOffFarm: OnOffFarm.fromState(state?.onOffFarm),
      cphNumber: CphNumber.fromState(state?.cphNumber),
      address: Address.fromState(state?.address)
    })
  }
}
