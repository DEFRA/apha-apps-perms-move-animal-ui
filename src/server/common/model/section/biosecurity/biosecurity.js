import { TbSectionModel } from '../../../../tb/section-model.js'
import { biosecurity } from '~/src/server/tb/biosecurity/index.js'
import { biosecuritySectionIsVisible } from '../visibility.js'
import { obligationsPage } from '~/src/server/tb/biosecurity/obligations/index.js'

/** @import {SectionConfig} from '~/src/server/common/model/section/section-model/section-model.js' */

export class BiosecuritySection extends TbSectionModel {
  /** @type {SectionConfig} */
  static config = {
    key: 'biosecurity',
    title: 'Biosecurity details',
    plugin: biosecurity,
    summaryLink: '/biosecurity/check-answers',
    isEnabled: () => true,
    isVisible: biosecuritySectionIsVisible
  }

  static firstPageFactory = () => obligationsPage
}
