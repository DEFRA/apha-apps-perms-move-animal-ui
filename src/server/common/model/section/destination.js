import { SectionModel } from '../section/section-model/index.js'
import { destinationTypePage } from '~/src/server/destination/destination-type/index.js'

/**
 * export @typedef {{
 * destinationType: DestinationTypeData | undefined;
 * }} DestinationData
 * @import {DestinationTypeData} from '../answer/destination-type/destination-type.js'
 */
export class Destination extends SectionModel {
  firstPage = destinationTypePage

  get destinationType() {
    return this._data?.destinationType.answer
  }

  /**
   * @param {DestinationData | undefined} state
   * @returns {Destination}
   */
  static fromState(state) {
    return new Destination({
      destinationType: {
        page: destinationTypePage,
        answer: destinationTypePage.Answer.fromState(state?.destinationType)
      }
    })
  }
}
