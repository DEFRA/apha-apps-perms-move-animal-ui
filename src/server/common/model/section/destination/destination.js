import { SectionModel } from '../section-model/section-model.js'

export class DestinationSection extends SectionModel {
  validate() {
    return { isValid: false, result: {} }
  }

  /**
   * @returns {DestinationSection}
   */
  static fromState() {
    return new DestinationSection()
  }
}
