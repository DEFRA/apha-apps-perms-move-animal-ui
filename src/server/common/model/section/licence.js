import { SectionModel } from '../section/section-model/index.js'
import { emailAddressPage } from '~/src/server/licence/email-address/index.js'

/**
 * export @typedef {{
 * emailAddress: EmailAddressData | undefined;
 * }} LicenceData
 * @import {EmailAddressData} from '../answer/email/email-address.js'
 */

export class Licence extends SectionModel {
  firstPage = emailAddressPage

  get emailAddress() {
    return this._data?.emailAddress.answer
  }

  /**
   * @param {LicenceData | undefined} state
   * @returns {Licence}
   */
  static fromState(state) {
    return new Licence({
      emailAddress: {
        page: emailAddressPage,
        answer: emailAddressPage.Answer.fromState(state?.emailAddress)
      }
    })
  }
}
