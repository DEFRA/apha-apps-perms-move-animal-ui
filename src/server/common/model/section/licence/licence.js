import { licence } from '~/src/server/licence/index.js'
import { SectionModel } from '../section-model/section-model.js'
import { fullNamePage } from '~/src/server/licence/fullName/index.js'

/** @import {SectionConfig} from '../section-model/section-model.js' */

export class LicenceSection extends SectionModel {
  /** @type {SectionConfig} */
  static config = {
    key: 'licence',
    title: 'Receiving the licence',
    plugin: licence,
    summaryLink: '/receiving-the-licence/check-answers',
    isEnabled: () => true,
    isVisible: () => true
  }

  static firstPageFactory = () => fullNamePage
}
