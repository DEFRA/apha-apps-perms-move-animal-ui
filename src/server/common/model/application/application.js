import { DestinationSection } from '../section/destination/destination.js'
import { LicenceSection } from '../section/licence/licence.js'
import { OriginSection } from '../section/origin/origin.js'
import { validateApplication } from './validation.js'

/**
 * export @typedef {{
 * origin: OriginData | undefined;
 * licence: LicenceData | undefined;
 * destination: DestinationData | undefined;
 * }} ApplicationData
 * @import {OriginData} from '../section/origin/origin.js'
 * @import {LicenceData} from '../section/licence/licence.js'
 * @import {DestinationData} from '../section/destination/destination.js'
 */

export class ApplicationModel {
  _data

  constructor(data) {
    this._data = data
    Object.seal(this)
  }

  validate() {
    return validateApplication(this._data)
  }

  /**
   * Getter for tasks.
   * @returns {Array|Object} The data associated with each and every task in the application
   */
  get tasks() {
    return this._data
  }

  /**
   * @param {ApplicationData | undefined} state
   * @returns {ApplicationModel}
   */
  static fromState(state) {
    return new ApplicationModel({
      origin: OriginSection.fromState(state?.origin),
      licence: LicenceSection.fromState(state?.licence),
      destination: DestinationSection.fromState(state?.destination)
    })
  }
}
