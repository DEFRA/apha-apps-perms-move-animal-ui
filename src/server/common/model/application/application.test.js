import { ApplicationModel } from './application.js'
import { LicenceSection } from '../section/licence/licence.js'
import { OriginSection } from '../section/origin/origin.js'
import { DestinationSection } from '../section/destination/destination.js'
import { BiosecuritySection } from '../section/biosecurity/biosecurity.js'
import { validApplicationState } from '../../test-helpers/journey-state.js'

describe('Application.fromState', () => {
  it('should create an Application instance from a valid state', () => {
    const application = ApplicationModel.fromState(validApplicationState)

    expect(application).toBeInstanceOf(ApplicationModel)

    expect(application.tasks.origin).toBeInstanceOf(OriginSection)
    expect(application.tasks.licence).toBeInstanceOf(LicenceSection)
    expect(application.tasks.identification).toBeUndefined()
    expect(application.tasks.destination).toBeInstanceOf(DestinationSection)
    expect(application.tasks.biosecurity).toBeInstanceOf(BiosecuritySection)

    expect(application.tasks.origin.questionPageAnswers).toEqual(
      OriginSection.fromState(validApplicationState).questionPageAnswers
    )
    expect(application.tasks.licence.questionPageAnswers).toEqual(
      LicenceSection.fromState(validApplicationState).questionPageAnswers
    )
    expect(application.tasks.destination.questionPageAnswers).toEqual(
      DestinationSection.fromState(validApplicationState).questionPageAnswers
    )
    expect(application.tasks.biosecurity.questionPageAnswers).toEqual(
      BiosecuritySection.fromState(validApplicationState).questionPageAnswers
    )
  })
})
