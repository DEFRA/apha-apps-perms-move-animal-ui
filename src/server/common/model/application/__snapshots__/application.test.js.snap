// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application.applicationData should render application data in the datastructure expected 1`] = `
{
  "journeyId": "TEST_APPLICATION",
  "journeyVersion": {
    "major": 1,
    "minor": 1,
  },
  "sections": [
    {
      "questionAnswers": [
        {
          "answer": {
            "displayText": "On to the farm or premises",
            "type": "text",
            "value": "on",
          },
          "question": "Are you moving the animals on or off your farm or premises?",
          "questionKey": "onOffFarm",
        },
        {
          "answer": {
            "displayText": "TB restricted farm",
            "type": "text",
            "value": "tb-restricted-farm",
          },
          "question": "Which type of premises are the animals moving off?",
          "questionKey": "originType",
        },
        {
          "answer": {
            "displayText": "12/345/6789",
            "type": "text",
            "value": "12/345/6789",
          },
          "question": "What is the county parish holding (CPH) number of the farm or premises where the animals are moving off?",
          "questionKey": "cphNumber",
        },
        {
          "answer": {
            "displayText": "Starfleet Headquarters
San Francisco
RG24 8RR",
            "type": "address",
            "value": {
              "addressLine1": "Starfleet Headquarters",
              "addressPostcode": "RG24 8RR",
              "addressTown": "San Francisco",
            },
          },
          "question": "What is the address of the farm or premises where the animals are moving off?",
          "questionKey": "address",
        },
      ],
      "sectionKey": "origin",
      "title": "Movement origin",
    },
    {
      "questionAnswers": [
        {
          "answer": {
            "displayText": "TB restricted farm",
            "type": "text",
            "value": "tb-restricted-farm",
          },
          "question": "Where are the animals going to?",
          "questionKey": "destinationType",
        },
        {
          "answer": {
            "displayText": "12/345/6789",
            "type": "text",
            "value": "12/345/6789",
          },
          "question": "What is the county parish holding (CPH) number of the farm or premises where the animals are going to?",
          "questionKey": "destinationFarmCph",
        },
        {
          "answer": {
            "displayText": "Starfleet Headquarters
San Francisco
RG24 8RR",
            "type": "address",
            "value": {
              "addressLine1": "Starfleet Headquarters",
              "addressPostcode": "RG24 8RR",
              "addressTown": "San Francisco",
            },
          },
          "question": "What is the address of the farm or premises where the animals are going to?",
          "questionKey": "destinationFarmAddress",
        },
        {
          "answer": {
            "displayText": "1",
            "type": "text",
            "value": "1",
          },
          "question": "How many animals are you planning to move?",
          "questionKey": "howManyAnimals",
        },
        {
          "answer": {
            "displayText": "Welfare",
            "type": "text",
            "value": "welfare",
          },
          "question": "What is the reason for the movement?",
          "questionKey": "reasonForMovement",
        },
        {
          "answer": {
            "displayText": "",
            "type": "text",
            "value": "",
          },
          "question": "Enter any additional information (optional)",
          "questionKey": "additionalInfo",
        },
      ],
      "sectionKey": "destination",
      "title": "Movement destination",
    },
  ],
}
`;
